### PR TITLE
Fix docstrings

### PR DESCRIPTION
**Pull request by erobitschek**
Addition and consistent formatting of docstrings (google format)

### CODE REVIEW

File: src/configs/config_scaffold.py [hunk 1]
```diff
@@ -66,8 +67,12 @@ class DatasetConfig:
...
     feature_threshold: int = 0 # the minimum number of times a medical code occurs in the dataset
     shuffle: bool = True
-    raw_dir: str = "../data/raw/"
-    processed_dir: str = "../data/processed/"
+
+    def get_data_paths(self):
+        raw_dir = f"../data/{self.project}/raw/"
+        processed_dir = f"../data/{self.project}/processed/"
+        split_dir = f"../data/{self.project}/split/"
+        return raw_dir, processed_dir, split_dir
```
Comment (Simsso) [inline src/configs/config_scaffold.py:75]: Instead you can add something called a property method. 

```python
@property
def raw_dir(self):
  return f"../data/{self.project}/raw/"
```

and access it with `config.raw_dir`. Or, maybe even better, set the value in the `__post_init__` function.

File: src/data.py [hunk 2]
```diff
@@ -59,19 +59,29 @@ def __getitem__(self, idx):
...
+        path (str): Path to the CSV file.
+        filter_cols (list, optional): List of columns to retain. Defaults to None.
+
+    Returns:
+        pd.DataFrame: Loaded data.
+    """
     if filter_cols:
         return pd.read_csv(path)[filter_cols]
     else:
         return pd.read_csv(path)
```
Comment (Simsso) [inline src/data.py:74]: Can be nicely shortened using a "unary operator".

File: src/data.py [hunk 3]
```diff
@@ -159,20 +176,30 @@ def get_x_y(
...
         print("Encoding feature counts.")
         features = encode_codes_with_counts(filtered_data).astype(int)
-    else: 
+    else:
         raise ValueError(f"Encoding method {encoding} not supported.")
 
-    assert features.index.equals(target_df.index), 'Feature and target var indices must match.'
+    assert features.index.equals(
+        target_df.index
+    ), "Feature and target var indices must match."
```
Comment (Simsso) [inline src/data.py:184]: Turn it into an if with raise ValueError. Assertions are only meant for expressing something that will always be true. Not for checking something.

File: src/eval.py [hunk 4]
```diff
@@ -1,22 +1,29 @@
...
 from data import DataSplit
 from predict import save_predictions_to_file
 
 
 def evaluate_predictions(
     predictions: npt.ArrayLike, true_labels: npt.ArrayLike
 ) -> Dict[str, float]:
-    """Evaluate predictions using various metrics."""
+    """
+    Evaluate predictions using various metrics such as AUC, balanced accuracy, F1 score, precision, and recall.
```
Comment (Simsso) [inline src/eval.py:18]: Now newline needed here.

`"""Evaluate ...` (no  space after `"""`)

File: src/models.py [hunk 5]
```diff
@@ -6,28 +6,17 @@ class TorchLogisticRegression(nn.Module):
...
-    input_dim : int
-        Dimensionality (i.e., number of features) of the input data.
+    Attributes:
+        linear (nn.Linear): A linear layer that transforms the input data.
+        input_dim (int): Dimensionality (i.e., number of features) of the input data.
+
+    Example:
+        >>> model = LogisticRegression(input_dim=10)
+        >>> sample_input = torch.rand((5, 10))
+        >>> output = model(sample_input)
```
Comment (Simsso) [inline src/models.py:16]: Nice!

File: src/models.py [hunk 6]
```diff
@@ -40,19 +29,21 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            torch.Tensor: Output tensor after passing through the linear layer
+                          and the sigmoid activation. Shape: (batch_size, 1).
         """
         out = torch.sigmoid(self.linear(x))
         return out
 
     def attributes(self):
-        return {"input_dim": self.input_dim}
+        """
+        Returns the attributes of the model as a dictionary.
```
Comment (Simsso) [inline src/models.py:44]: No newline needed. Check all occurrences, ideally.

Comment (Simsso) [reply src/models.py:44]: In general, for this fn a single-line docstring is sufficient.

File: src/run_analysis.py [hunk 7]
```diff
@@ -13,6 +13,20 @@
 
 
 def parse_args():
+    """
+    Parses command-line arguments for the script using argparse.
+    
+    Args:
+        None
```
Comment (Simsso) [inline src/run_analysis.py:20]: Just remove the `Args:` section in this case.

File: src/run_analysis.py [hunk 8]
```diff
@@ -13,6 +13,20 @@
...
+
+    Returns:
+        argparse.Namespace: Parsed command-line arguments.
+
+    Notes:
+        This function sets up the arguments that the main script expects, including configuration 
+        paths, training modes, and evaluation settings. Make sure to provide the necessary arguments 
+        when running the script.
```
Comment (Simsso) [inline src/run_analysis.py:28]: Such notes go underneath the first line of the comment. The docstring should be like this:

```
    """Parses command-line arguments for the script using argparse.

    This function sets up the arguments that the main script expects, including configuration 
    paths, training modes, and evaluation settings. Make sure to provide the necessary arguments 
    when running the script.

    Returns:
        argparse.Namespace: Parsed command-line arguments.
    """
```

File: src/run_analysis.py [hunk 9]
```diff
@@ -38,6 +52,26 @@ def parse_args():
...
 
 def main():
+    """
+    Main function to run the model processing pipeline based on command-line arguments.
+
+    Args:
+        None
+
+    Returns:
+        None
```
Comment (Simsso) [inline src/run_analysis.py:62]: Remove these

File: src/run_analysis.py [hunk 10]
```diff
@@ -38,6 +52,26 @@ def parse_args():
...
+
+    Raises:
+        ValueError: If an unsupported train mode is provided.
+        ValueError: If an unsupported data state is provided.
+        ValueError: If an unsupported model framework is provided.
```
Comment (Simsso) [inline src/run_analysis.py:67]: Collapse into one line: If an unsupported xx / yy / zz is provided. Generally, you don't need to mention that explicitly, if it's a function that is not called by anyone (as this one).

File: src/utils.py [hunk 11]
```diff
@@ -43,6 +41,12 @@ def set_seed():
 
 
 def remove_dir_contents(dir_path: str) -> None:
+    """
+    Removes all contents (files and sub-directories) from a specified directory.
+
+    Args:
+        dir_path (str): Path to the directory from which contents will be removed.
```
Comment (Simsso) [inline src/utils.py:48]: Can be trivially inferred from the function signature.

**Review by Simsso — COMMENTED** [review]
Nice! Couple of comments which apply to all docstrings. After addressing these you're good to go and can merge the PR.

File: src/configs/config_scaffold.py (hunk 1 above)
Comment (erobitschek) [reply src/configs/config_scaffold.py:75]: WHOAAAA

Comment (erobitschek) [reply src/configs/config_scaffold.py:75]: SO. COOL. :D

File: src/configs/config_scaffold.py [hunk 12]
```diff
@@ -58,6 +58,7 @@ class ModelConfig:
 @dataclass
```
[the remark this answers is not in the document]
Comment (erobitschek) [reply src/configs/config_scaffold.py:58]: Oh I think I came across this before and it has to do with the immutability of the data, right? :) I'll look into it more.

File: src/data.py [hunk 13]
```diff
@@ -59,19 +59,29 @@ def __getitem__(self, idx):
 
 
 def load_data(path: str, filter_cols: list = None) -> pd.DataFrame:
+    """Loads data from a CSV file and optionally filters certain columns.
+
+    Args:
+        path (str): Path to the CSV file.
+        filter_cols (list, optional): List of columns to retain. Defaults to None.
```
[the remark this answers is not in the document]
Comment (erobitschek) [reply src/data.py:66]: will remove those

File: src/data.py (hunk 3 above)
Comment (erobitschek) [reply src/data.py:182]: Thanks for the reminder - fixed

File: src/data.py (hunk 2 above)
Comment (erobitschek) [reply src/data.py:67]: got it :D

File: src/configs/config_scaffold.py (hunk 1 above)
Comment (erobitschek) [reply src/configs/config_scaffold.py:75]: I changed the implementation to a __post_init__ call. When would you recommend using property methods versus post init?

**Comment by Simsso** [comment]
You can delete the branch :-)

**Pull request merged**